### DISCUSSION TITLE
Dynamic routes generation and pages creation based on templates

### DISCUSSION THREAD

**Discussion by lnikell** [vercel/next.js#20377] [Help]
Hi, 

I'm considering moving from Gatsby to Next.js for websites that we build for our clients. Usually, our setup is Gatsby + WordPress, but it does not work well with even medium size website - from 50 pages. The dev experience on local machine is not great because you have to fetch whole data from WP, then generate thumbnails, etc. So builds take up to 10 minutes without cache, and even with cached versions, it takes 2-3 minutes. 

I understand how can I replace most of the part with Next.js, however, I don't understand how can I dynamically generate/refresh routes based on pages and URLs that were created in WP.

Example:
Client creates in WP new page "/about" and want to use a template called "About" that we made specifically for that page. Then Client can creates sub page to that parent page like /about/team and use the template "Static page" for this. 

I want to have different templates that will be placed inside a templates directory, then there will be sort of a function that I will be able to re-run based on changes in WP. Inside that function, there will be a way of programmatically re-map specific routes to the specific templates(component). Those components should have getServerSideProps in order to get data required for those templates.

**Comment by kelvinlouis** [comment]
Hey Inikell,

We solved this issue in two different ways so far in our projects.

**Custom server**
Before there were [dynamic routes](https://nextjs.org/docs/routing/dynamic-routes) we used a [custom server](https://nextjs.org/docs/advanced-features/custom-server) to do the routing in the backend. This is not very optimal (see info boxes in the linked page) because you lose access to certain features of Next.js. However, this gives you full control to do routing and more.

**Dynamic routes**
You have the option to create [dynamic routes](https://nextjs.org/docs/routing/dynamic-routes). I think you will be interested in the catch all routes feature, but this depends how flexible/dynamic your routing of the website has to be.
You could for example define `/pages/about/index.js` for your about page. For the sub pages you can use a catch all route `/pages/about/[...slug].js`. If you only render default pages for these sub-routes, you could render `/components/DefaultPage.js` in them. I recommend to create a reusable component, especially if other routes can render default pages. If sub pages could be of any type, I recommend that your WP instance returns the page type as an attribute when fetching data. Basically you fetch the data in `getServerSideProps` and pass the fetched data to the component. Depending on the type you can render a different page component. The component of `/pages/about/[...slug].js` page would act as a factory. Depending on the type you could render a different page component. I have never tried this, but you could try to do: `/pages/[[...slug]].js` and have your entire website be dynamic.

Hope this helps.

**Reply by lnikell** [reply]
Thank you for your answer! I'm considering at this moment using this approach https://github.com/agility/agilitycms-next-starter-ssg/tree/master/pages since I want to make sure that the URL structure is driven by cms. Not sure yet how to organize everything better, but I think some ideas will come up.
